File: dailyGroupWars/June24/138.cpp
#include <unordered_map>

class Node {
public:
  int val;
  Node *next;
  Node *random;

  Node(int _val) {
    val = _val;
    next = nullptr;
    random = nullptr;
  }
};

class Solution {
public:
  Node *copyRandomList(Node *head) {
    std::unordered_map<Node *, Node *> visited;
    return dfs(head, visited);
  }

  Node *dfs(Node *current, std::unordered_map<Node *, Node *> &visited) {
    if (!current) {
      return nullptr;
    }
    if (visited.find(current) != visited.end()) {
      return visited[current];
    }
    Node *curr = new Node(current->val);
    visited[current] = curr;
    curr->next = dfs(current->next, visited);
    curr->random = dfs(current->random, visited);
    return curr;
  }
};

File: src/150redo/LinkedList/138.go
package LinkedList

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

var (
	mp = make(map[*Node]*Node)
)

// look to perform deeop copy of this list
func copyRandomList(head *Node) *Node {
	return f(head)
}

func f(curr *Node) *Node {
	if value, ok := mp[curr]; ok {
		return value
	}

	if curr == nil {
		return nil
	}
	// make copy of this Node
	cpy := &Node{}
	mp[curr] = cpy // THIS IS IMPORTANT => Not doing this here can reuslt in infinite recursion, need this to so when reaching here again, we return instead of creating another instance
	cpy.Val = curr.Val
	cpy.Next = f(curr.Next)
	cpy.Random = f(curr.Random)

	return cpy
}
